File: sdk/python/rexec/terminal.py
"""Terminal service for WebSocket connections to containers."""

import asyncio
import json
from typing import TYPE_CHECKING, AsyncIterator, Optional

import websockets
from websockets.client import WebSocketClientProtocol

from rexec.exceptions import RexecConnectionError

if TYPE_CHECKING:
    from rexec.client import RexecClient

__all__ = ["TerminalService", "Terminal"]


class Terminal:
    """
    WebSocket terminal connection to a container.

    Use as an async context manager or manually manage the connection.

    Example:
        async with client.terminal.connect(container.id) as term:
            await term.write(b"ls -la\\n")
            async for data in term:
                print(data.decode(), end="")
    """

    def __init__(self, ws: WebSocketClientProtocol):
        self._ws = ws
        self._closed = False

    async def write(self, data: bytes | str) -> None:
        """
        Send data to the terminal.

        Args:
            data: Data to send (bytes or string).
        """
        if self._closed:
            raise RexecConnectionError("Terminal connection is closed")

        if isinstance(data, str):
            data = data.encode()
        await self._ws.send(data)

    async def read(self) -> bytes:
        """
        Read data from the terminal.

        Returns:
            Bytes received from the terminal.

        Raises:
            RexecConnectionError: If connection is closed.
        """
        if self._closed:
            raise RexecConnectionError("Terminal connection is closed")

        try:
            data = await self._ws.recv()
            if isinstance(data, str):
                return data.encode()
            return data
        except websockets.ConnectionClosed:
            self._closed = True
            raise RexecConnectionError("Terminal connection closed")

    async def resize(self, cols: int, rows: int) -> None:
        """
        Resize the terminal.

        Args:
            cols: Number of columns.
            rows: Number of rows.
        """
        msg = json.dumps({"type": "resize", "cols": cols, "rows": rows})
        await self._ws.send(msg)

    async def close(self) -> None:
        """Close the terminal connection."""
        if not self._closed:
            self._closed = True
            await self._ws.close()

    @property
    def closed(self) -> bool:
        """Check if the connection is closed."""
        return self._closed

    def __aiter__(self) -> AsyncIterator[bytes]:
        """Iterate over terminal output."""
        return self

    async def __anext__(self) -> bytes:
        """Get next chunk of terminal output."""
        try:
            return await self.read()
        except RexecConnectionError:
            raise StopAsyncIteration

    async def __aenter__(self) -> "Terminal":
        """Enter async context."""
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        """Exit async context."""
        await self.close()


class TerminalService:
    """Service for terminal WebSocket connections."""

    def __init__(self, client: "RexecClient"):
        self._client = client

    async def connect(
        self,
        container_id: str,
        *,
        cols: int = 80,
        rows: int = 24,
        timeout: float = 30.0,
    ) -> Terminal:
        """
        Connect to a container's terminal.

        Args:
            container_id: The container ID.
            cols: Terminal width in columns (default: 80).
            rows: Terminal height in rows (default: 24).
            timeout: Connection timeout in seconds (default: 30).

        Returns:
            Terminal object for reading/writing.

        Example:
            term = await client.terminal.connect(container.id)
            try:
                await term.write(b"echo hello\\n")
                data = await term.read()
                print(data.decode())
            finally:
                await term.close()
        """
        ws_url = self._client._ws_url(f"/ws/terminal/{container_id}")

        try:
            ws = await asyncio.wait_for(
                websockets.connect(
                    ws_url,
                    additional_headers={"Authorization": f"Bearer {self._client._token}"},
                ),
                timeout=timeout,
            )
        except asyncio.TimeoutError:
            raise RexecConnectionError(f"Connection timeout after {timeout}s")
        except Exception as e:
            raise RexecConnectionError(f"Failed to connect: {e}")

        terminal = Terminal(ws)

        # Set initial size
        if cols and rows:
            await terminal.resize(cols, rows)

        return terminal

File: sdk/ruby/lib/rexec/terminal.rb
# frozen_string_literal: true

require "json"

module Rexec
  # WebSocket terminal connection.
  class Terminal
    attr_reader :closed

    def initialize(ws)
      @ws = ws
      @closed = false
      @data_handlers = []
      @close_handlers = []
      @error_handlers = []

      setup_handlers
    end

    # Send data to the terminal.
    #
    # @param data [String] Data to send
    def write(data)
      raise TerminalClosedError if @closed

      @ws.send(data)
    end

    # Resize the terminal.
    #
    # @param cols [Integer] Number of columns
    # @param rows [Integer] Number of rows
    def resize(cols, rows)
      raise TerminalClosedError if @closed

      msg = { type: "resize", cols: cols, rows: rows }.to_json
      @ws.send(msg)
    end

    # Register a handler for incoming data.
    #
    # @yield [data] Block called with each chunk of data
    def on_data(&block)
      @data_handlers << block
    end

    # Register a handler for connection close.
    #
    # @yield Block called when connection closes
    def on_close(&block)
      @close_handlers << block
    end

    # Register a handler for errors.
    #
    # @yield [error] Block called with error
    def on_error(&block)
      @error_handlers << block
    end

    # Close the terminal connection.
    def close
      return if @closed

      @closed = true
      @ws.close
    end

    alias closed? closed

    private

    def setup_handlers
      @ws.on :message do |msg|
        @data_handlers.each { |h| h.call(msg.data) }
      end

      @ws.on :close do
        @closed = true
        @close_handlers.each(&:call)
      end

      @ws.on :error do |e|
        @error_handlers.each { |h| h.call(e) }
      end
    end
  end

  # Service for terminal WebSocket connections.
  class TerminalService
    def initialize(client)
      @client = client
    end

    # Connect to a container's terminal.
    #
    # @param container_id [String] Container ID
    # @param cols [Integer] Terminal width (default: 80)
    # @param rows [Integer] Terminal height (default: 24)
    # @return [Terminal]
    #
    # @example
    #   terminal = client.terminal.connect(container.id)
    #   terminal.on_data { |data| print data }
    #   terminal.write("ls -la\n")
    def connect(container_id, cols: 80, rows: 24)
      begin
        require "websocket-client-simple"
      rescue LoadError
        raise LoadError, "websocket-client-simple is required for terminal connections (gem install websocket-client-simple)"
      end
      ws_url = @client.ws_url("/ws/terminal/#{container_id}")

      ws = WebSocket::Client::Simple.connect(ws_url, headers: {
        "Authorization" => "Bearer #{@client.token}"
      })

      terminal = Terminal.new(ws)

      # Wait for connection
      sleep 0.1 until ws.open?

      # Set initial size
      terminal.resize(cols, rows)

      terminal
    end
  end
end
